BoxContentSDK/BoxContentSDK/OAuth2/BOXParallelOAuth2Session.m

```diff
@@ -44,19 +45,44 @@ - (void)performRefreshTokenGrant:(NSString *)expiredAccessToken withCompletionBl
         {
             // Only attempt to refresh the token if this is the first time this access
             // token has expired
+            
+            // TODO We need to ensure that completion block gets called, but we don't want to trigger duplicate attempts to refresh the same token.
+            // We should have an observer-notifier pattern to hook up this block into the existing refresh operation in progress.
+            // Without that, the best we can do is return this error.
+            if (block) {
+                block(self, [NSError errorWithDomain:BOXContentSDKErrorDomain code:BOXContentSDKAuthErrorTokenRefreshAlreadyInProgress userInfo:nil]);
+            }
             return;
         }
         
-        BOXLog(@"access token expired: %@", expiredAccessToken);
-        BOXLog(@"refreshing tokens");
         if (expiredAccessToken)
         {
             [self.expiredOAuth2Tokens addObject:expiredAccessToken];
         }
         
-        [super performRefreshTokenGrant:expiredAccessToken withCompletionBlock:block];
+        [super performRefreshTokenGrant:expiredAccessToken withCompletionBlock:^(BOXAbstractSession *session, NSError *error) {
+            if (error != nil && ![self isInvalidGrantError:error]) {
+                // If there was an error, remove from 'expiredOAuth2Tokens' so that we can try again. For example, if there was a network
+                // error, then we would not want to block ourselves from trying to refresh the tokens again later.
+                // The exception to this is if we get 'invalid_grant' in which case the token is not just expired, but is just dead. In that
+                // case, we don't want to ever try refreshing that again.
+                [self.expiredOAuth2Tokens removeObject:expiredAccessToken];
+            }
+        }];
     }
 }
 
+- (BOOL)isInvalidGrantError:(NSError *)error
+{
+    BOOL isInvalidGrantError = NO;
+    if ([error.domain isEqualToString:BOXContentSDKErrorDomain]) {
```

Comment:
We don't have this already somewhere we can reuse? Can any other code piggy-back on this now?